staging/src/k8s.io/apiextensions-apiserver/pkg/storageversion/manager.go

```diff
@@ -0,0 +1,418 @@
+// CRD and returns immediately. Optionally, the caller may specify a
+// non-nil waitCh and/or a non-nil processedCh.
+// A non-nil waitCh will block the StorageVersion update until waitCh is
+// closed.
+// The manager will close the non-nil processedCh if it finished
+// processing the StorageVersion update (note that the update can either
+// succeeded or failed).
+func (m *Manager) updateStorageVersion(ctx context.Context, crd *apiextensionsv1.CustomResourceDefinition,
+	waitCh <-chan struct{}, teardownFinishedTimeout time.Duration) error {
+	if waitCh != nil {
+		done := false
+		for {
+			select {
+			case <-waitCh:
+				done = true
+			case <-ctx.Done():
+				return fmt.Errorf("aborting storageversion update for %v, context closed", crd)
+			case <-time.After(teardownFinishedTimeout):
+				return fmt.Errorf("timeout waiting for waitCh to close before proceeding with storageversion update for %v", crd)
+			}
+			if done {
+				break
+			}
+		}
+	}
+
+	if err := m.updateCRDStorageVersion(ctx, crd); err != nil {
+		return fmt.Errorf("error while updating storage version for crd %v: %w", crd, err)
+	}
+
+	return nil
+}
+
+func (m *Manager) updateCRDStorageVersion(ctx context.Context, crd *apiextensionsv1.CustomResourceDefinition) error {
+	gr := schema.GroupResource{
+		Group:    crd.Spec.Group,
+		Resource: crd.Spec.Names.Plural,
+	}
+	storageVersion, err := apiextensionshelpers.GetCRDStorageVersion(crd)
+
+	if err != nil {
+		// This should never happen if crd is valid, which is true since we
+		// only update storage version for CRDs that have been written to the
+		// storage.
+		return err
+	}
+
+	encodingVersion := crd.Spec.Group + "/" + storageVersion
+	var servedVersions, decodableVersions []string
+	for _, v := range crd.Spec.Versions {
+		decodableVersions = append(decodableVersions, crd.Spec.Group+"/"+v.Name)
+		if v.Served {
+			servedVersions = append(servedVersions, crd.Spec.Group+"/"+v.Name)
+		}
+	}
+
+	appendOwnerRefFunc := func(sv *apiserverinternalv1alpha1.StorageVersion) error {
+		ref := metav1.OwnerReference{
+			APIVersion: apiextensionsv1.SchemeGroupVersion.String(),
+			Kind:       "CustomResourceDefinition",
+			Name:       crd.Name,
+			UID:        crd.UID,
+		}
+		for _, r := range sv.OwnerReferences {
+			if r == ref {
+				return nil
+			}
+		}
+		sv.OwnerReferences = append(sv.OwnerReferences, ref)
+		return nil
+	}
+	return genericstorageversion.UpdateStorageVersionFor(
+		ctx,
+		m.client,
+		m.apiserverID,
+		gr,
+		encodingVersion,
+		decodableVersions,
+		servedVersions,
+		appendOwnerRefFunc)
+}
+
+func (m *Manager) processStorageVersionUpdateResponse(err error, svUpdateInfo *storageVersionUpdateInfo) {
+	if err == nil {
+		klog.V(4).Infof("successfully updated storage version for %s", svUpdateInfo.crd.Name)
+		close(svUpdateInfo.updateChannels.processedCh)
+		return
+	}
+
+	failuresSeenSoFar := svUpdateInfo.failures + 1
+	if failuresSeenSoFar > storageversionUpdateFailureThreshold {
```

Comment:
Can a CRD handler become indefinitely wedged if storageversionUpdateFailureThreshold is exceeded?

I'm imagining a case where, say, there is an etcd outage that starts right around the first sv update, and the threshold is exceeded before etcd becomes available again.  Ideally, once etcd recovers, the storage version update would eventually also be completed and the CRD handler would be updated.

It looks like if we don't continue to enqueue here that the only way to recover from my scenario would be to make a spurious update ot the CRD to force Enqueue?